nearcore/src/state_sync.rs

```diff
@@ -360,17 +378,26 @@ fn start_dumping(
     sync_hash: CryptoHash,
     shard_id: ShardId,
     chain: &Chain,
-    runtime: &Arc<NightshadeRuntime>,
+    runtime: &Arc<dyn RuntimeWithEpochManagerAdapter>,
+    account_id: &Option<AccountId>,
 ) -> Result<Option<StateSyncDumpProgress>, Error> {
     let epoch_info = runtime.get_epoch_info(&epoch_id)?;
     let epoch_height = epoch_info.epoch_height();
-    let sync_prev_header = chain.get_block_header(&sync_hash)?;
-    let sync_prev_hash = sync_prev_header.hash();
+
+    let sync_header = chain.get_block_header(&sync_hash)?;
+    let sync_prev_hash = sync_header.prev_hash();
+    let sync_prev_header = chain.get_block_header(&sync_prev_hash)?;
+    // Need to check if the completed epoch had a shard this account cares about.
+    // sync_hash is the first block of the next epoch.
+    // `cares_about_shard()` accepts `parent_hash`, therefore we need prev-prev-hash,
+    // and its next-hash will be prev-hash. That is the last block of the completed epoch,
+    // which is what we wanted.
```

Comment:
related to this, it looks to me like we are passing the wrong `block_hash` to `runtime.obtain_state_part` in `obtain_and_store_state_part()`. If you look at how it's done in `Chain::get_state_response_part()`, it should be the prev hash before `sync_hash`, not `sync_hash` itself

Reply:
Yes, thank you for noticing.
Fixed.